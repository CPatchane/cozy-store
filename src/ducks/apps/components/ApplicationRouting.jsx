import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import IntentModal from 'cozy-ui/react/IntentModal'

import UninstallModal from './UninstallModal'
import InstallModal from './InstallModal'
import ApplicationPage from './ApplicationPage'

import { APP_TYPE } from 'ducks/apps'

export class ApplicationRouting extends Component {
  render () {
    const {
      apps,
      installedApps,
      isFetching,
      isInstalling,
      parent,
      history,
      actionError,
      fetchError
    } = this.props
    const appsArray = apps || installedApps || []
    return (
      <div>
        <Route
          path={`/${parent}/:appSlug`}
          render={({ match }) => {
            if (isFetching) return
            if (appsArray.length && match.params) {
              const app = appsArray.find(
                app => app.slug === match.params.appSlug
              )
              if (!app) return history.push(`/${parent}`)
              return <ApplicationPage app={app} parent={parent} />
            }
          }}
        />
        <Route
          path={`/${parent}/:appSlug/manage`}
          render={({ match }) => {
            if (isFetching) return
            if (appsArray.length && match.params) {
              const app = appsArray.find(
                app => app.slug === match.params.appSlug
              )
              if (!app) return history.push(`/${parent}`)
              if (app && app.installed) {
                return (
                  <UninstallModal
                    uninstallApp={this.props.uninstallApp}
                    parent={`/${parent}`}
                    uninstallError={actionError}
                    app={app}
                  />
                )
              } else {
                return (
                  <InstallModal
                    installApp={this.props.installApp}
                    parent={`/${parent}`}
                    installError={actionError}
                    fetchError={fetchError}
                    app={app}
                    isInstalling={isInstalling}
                  />
                )
              }
            }
          }}
        />
        <Route
          path={`/${parent}/:appSlug/configure`}
          render={({ match }) => {
            if (isFetching) return
            if (appsArray.length && match.params) {
              const appSlug = match.params.appSlug
              const app = appsArray.find(
                app => app.slug === appSlug
              )
              if (!app) return history.push(`/${parent}`)
              const goToApp = () => history.push(`/${parent}/${appSlug}`)
              if (app && app.installed && app.type === APP_TYPE.KONNECTOR) {
                return (<IntentModal
                  action='CREATE'
                  doctype='io.cozy.accounts'
                  options={{ slug: appSlug }}
                  dismissAction={goToApp}
                  onComplete={goToApp}
                />)
              } else {
                return goToApp()
              }
            }
          }}
        />
      </div>
    )
  }
}

export default withRouter(ApplicationRouting)
