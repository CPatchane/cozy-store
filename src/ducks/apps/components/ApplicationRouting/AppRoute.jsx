import React from 'react'
import { Route } from 'react-router-dom'

import ApplicationPage from '../ApplicationPage'

const AppRoute = ({ parent, getApp, isFetching, redirectTo }) => (
  <Route
    path={`/${parent}/:appSlug`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      if (!app) return redirectTo(`/${parent}`)
      return <ApplicationPage app={app} parent={parent} />
    }}
  />
)

export default AppRoute
