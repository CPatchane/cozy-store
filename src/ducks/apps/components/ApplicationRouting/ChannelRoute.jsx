import React from 'react'
import { Route } from 'react-router-dom'

import Install from '../Install'
import { REGISTRY_CHANNELS } from 'ducks/apps'

export const ChannelRoute = ({
  actionError,
  fetchError,
  fetchLatestApp,
  getApp,
  installApp,
  isAppFetching,
  isFetching,
  isInstalling,
  parent,
  redirectTo,
  updateApp
}) => (
  <Route
    path={`/${parent}/:appSlug/channel/:channel`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      if (!app || !app.isInRegistry) return redirectTo(`/${parent}`)
      const channel = match.params.channel
      const isChannelAvailable = Object.values(REGISTRY_CHANNELS).includes(
        channel
      )
      if (!isChannelAvailable) {
        return redirectTo(`/${parent}/${app.slug}`)
      }
      const fetchApp = chan => fetchLatestApp(app.slug, chan)
      return (
        <Install
          installApp={app.installed ? updateApp : installApp}
          parent={`/${parent}`}
          fetchApp={fetchApp}
          isAppFetching={isAppFetching}
          installError={actionError}
          fetchError={fetchError}
          app={app}
          isInstalling={isInstalling}
          channel={channel}
        />
      )
    }}
  />
)

export default ChannelRoute
