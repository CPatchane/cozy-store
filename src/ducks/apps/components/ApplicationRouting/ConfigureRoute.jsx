import React from 'react'
import { Route } from 'react-router-dom'

import IntentModal from 'cozy-ui/react/IntentModal'
import { APP_TYPE } from 'ducks/apps'

const ConfigureRoute = ({ getApp, isFetching, parent, redirectTo }) => (
  <Route
    path={`/${parent}/:appSlug/configure`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      if (!app) return redirectTo(`/${parent}`)
      const goToApp = () => history.replace(`/${parent}/${app.slug}`)
      if (app && app.installed && app.type === APP_TYPE.KONNECTOR) {
        return (
          <IntentModal
            action="CREATE"
            doctype="io.cozy.accounts"
            options={{ slug: app.slug }}
            dismissAction={goToApp}
            onComplete={goToApp}
            mobileFullscreen
            overflowHidden
            size="small"
            height="35rem"
          />
        )
      } else {
        return goToApp()
      }
    }}
  />
)

export default ConfigureRoute
