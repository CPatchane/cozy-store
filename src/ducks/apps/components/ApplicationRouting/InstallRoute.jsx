import React from 'react'
import { Route } from 'react-router-dom'

import InstallModal from '../InstallModal'

export const InstallRoute = ({
  actionError,
  getApp,
  installApp,
  isFetching,
  isInstalling,
  parent,
  redirectTo,
  updateApp
}) => (
  <Route
    path={`/${parent}/:appSlug/install`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      if (!app || (app.installed && !app.availableVersion)) {
        // not existing or already installed with the
        // latest version
        return redirectTo(`/${parent}`)
      } else {
        return (
          <InstallModal
            installApp={app.installed ? updateApp : installApp}
            parent={`/${parent}`}
            installError={actionError}
            app={app}
            isInstalling={isInstalling}
          />
        )
      }
    }}
  />
)

export default InstallRoute
