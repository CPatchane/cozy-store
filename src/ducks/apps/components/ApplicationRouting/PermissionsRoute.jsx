import React from 'react'
import { Route } from 'react-router-dom'

import PermissionsModal from '../PermissionsModal'

export const PermissionsRoute = ({
  getApp,
  isFetching,
  parent,
  redirectTo
}) => (
  <Route
    path={`/${parent}/:appSlug/permissions`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      if (!app) return redirectTo(`/${parent}`)
      return <PermissionsModal app={app} parent={`/${parent}`} />
    }}
  />
)

export default PermissionsRoute
