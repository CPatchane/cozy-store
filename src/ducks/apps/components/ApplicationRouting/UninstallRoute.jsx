import React from 'react'
import { Route } from 'react-router-dom'

import UninstallModal from '../UninstallModal'

export const UninstallRoute = ({
  actionError,
  getApp,
  isFetching,
  isUninstalling,
  parent,
  uninstallApp
}) => (
  <Route
    path={`/${parent}/:appSlug/uninstall`}
    render={({ match }) => {
      if (isFetching) return
      const app = getApp(match)
      return (
        <UninstallModal
          uninstallApp={uninstallApp}
          isUninstalling={isUninstalling}
          parent={`/${parent}`}
          uninstallError={actionError}
          app={app}
        />
      )
    }}
  />
)

export default UninstallRoute
