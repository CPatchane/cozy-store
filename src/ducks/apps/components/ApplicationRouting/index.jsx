import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'

import AppRoute from './AppRoute'
import ChannelRoute from './ChannelRoute'
import PermissionsRoute from './PermissionsRoute'
import ConfigureRoute from './ConfigureRoute'
import InstallRoute from './InstallRoute'
import UninstallRoute from './UninstallRoute'

export class ApplicationRouting extends Component {
  componentWillReceiveProps(nextProps) {
    // don't react on error
    if (nextProps.actionError) return
    if (this.props.isUninstalling && !nextProps.isUninstalling) {
      const { location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return this.redirectTo(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      return this.redirectTo(`/${parent}/${app.slug}`)
    }
  }

  getAppFromMatchOrSlug = (match, slug) => {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (match && match.params && match.params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  redirectTo = target => {
    return this.props.history.replace(target)
  }

  render() {
    const {
      fetchLatestApp,
      installApp,
      uninstallApp,
      updateApp,
      isFetching,
      isAppFetching,
      isInstalling,
      isUninstalling,
      parent,
      actionError,
      fetchError
    } = this.props
    return (
      <div>
        <AppRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
        <ChannelRoute
          actionError={actionError}
          fetchError={fetchError}
          fetchLatestApp={fetchLatestApp}
          getApp={this.getAppFromMatchOrSlug}
          installApp={installApp}
          isAppFetching={isAppFetching}
          isFetching={isFetching}
          isInstalling={isInstalling}
          parent={parent}
          redirectTo={this.redirectTo}
          updateApp={updateApp}
        />
        <InstallRoute
          actionError={actionError}
          getApp={this.getAppFromMatchOrSlug}
          installApp={installApp}
          isFetching={isFetching}
          isInstalling={isInstalling}
          parent={parent}
          redirectTo={this.redirectTo}
          updateApp={updateApp}
        />
        <UninstallRoute
          actionError={actionError}
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          isUninstalling={isUninstalling}
          parent={parent}
          redirectTo={this.redirectTo}
          uninstallApp={uninstallApp}
        />
        <PermissionsRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
        <ConfigureRoute
          getApp={this.getAppFromMatchOrSlug}
          isFetching={isFetching}
          parent={parent}
          redirectTo={this.redirectTo}
        />
      </div>
    )
  }
}

export default withRouter(translate()(ApplicationRouting))
