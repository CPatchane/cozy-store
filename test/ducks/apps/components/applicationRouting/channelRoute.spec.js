'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import { Route } from 'react-router-dom'

import { ChannelRoute } from 'ducks/apps/components/ApplicationRouting/ChannelRoute'

import mockApps from '../../_mockApps'

const getAppMock = ({ params }) => mockApps.find(a => a.slug === params.appSlug)

const getProps = (isFetching = false, getApp = getAppMock) => ({
  fetchLatestApp: jest.fn(),
  isFetching,
  parent: 'myapps',
  getApp,
  redirectTo: jest.fn(() => null)
})

describe('ChannelRoute component', () => {
  it('should display install modal if app found and in the registry and installed', () => {
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    const route = component.find(Route).getElement()
    // collect in mockApps is installed and isInRegistry
    const routeProps = {
      match: { params: { appSlug: 'collect', channel: 'beta' } }
    }
    const resultComponent = route.props.render(routeProps)
    expect(resultComponent).toMatchSnapshot()
  })

  it('should display install modal if app found and in the registry even if not installed', () => {
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    const route = component.find(Route).getElement()
    // photos in mockApps is isInRegistry
    const routeProps = {
      match: { params: { appSlug: 'photos', channel: 'beta' } }
    }
    const resultComponent = route.props.render(routeProps)
    expect(resultComponent).toMatchSnapshot()
  })

  it('should display nothing if isFetching', () => {
    const props = getProps(true)
    const component = shallow(<ChannelRoute {...props} />)
    const route = component.find(Route).getElement()
    // collect in mockApps is installed and isInRegistry
    const routeProps = { match: { params: { appSlug: 'collect' } } }
    const resultComponent = route.props.render(routeProps)
    expect(resultComponent).toBe(null)
  })

  it('should redirectTo parent if no app found', () => {
    const props = getProps(false, jest.fn())
    const component = shallow(<ChannelRoute {...props} />)
    const route = component.find(Route).getElement()
    const routeProps = { match: { params: {} } }
    const resultComponent = route.props.render(routeProps)
    expect(props.redirectTo.mock.calls.length).toBe(1)
    expect(props.redirectTo.mock.calls[0][0]).toBe(`/${props.parent}`)
    expect(resultComponent).toBe(null)
  })

  it('should redirectTo app page if app channel not available', () => {
    const props = getProps()
    const component = shallow(<ChannelRoute {...props} />)
    const route = component.find(Route).getElement()
    // collect in mockApps is installed and isInRegistry
    const routeProps = {
      match: { params: { appSlug: 'collect', channel: 'jeanluc' } }
    }
    const resultComponent = route.props.render(routeProps)
    expect(props.redirectTo.mock.calls.length).toBe(1)
    expect(props.redirectTo.mock.calls[0][0]).toBe(
      `/${props.parent}/${routeProps.match.params.appSlug}`
    )
    expect(resultComponent).toBe(null)
  })
})
