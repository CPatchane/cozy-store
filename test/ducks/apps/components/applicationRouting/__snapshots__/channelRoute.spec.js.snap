// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelRoute component should display install modal if app found and in the registry and installed 1`] = `
<Connect(ChannelModal)
  appSlug="collect"
  channel="beta"
  dismissAction={[Function]}
  fetchApp={[Function]}
  onCurrentChannel={[Function]}
  onNotInstalled={[Function]}
  onSuccess={[Function]}
/>
`;

exports[`ChannelRoute component should display install modal if app found and in the registry even if not installed 1`] = `
<Connect(ChannelModal)
  appSlug="photos"
  channel="beta"
  dismissAction={[Function]}
  fetchApp={[Function]}
  onCurrentChannel={[Function]}
  onNotInstalled={[Function]}
  onSuccess={[Function]}
/>
`;
